README.md
Obsolete
========

With the advent of [Voctomix](https://github.com/voc/voctomix), this has become obsolete. Please head over there.

If you really want to use dvswitch-voc, check the history of this repository, which remains in place.


ChangeLog
dvswitch (0.8.3.6) unstable; urgency=low

  * Fix build failure with ffmpeg 0.7

 -- Ben Hutchings <ben@decadent.org.uk>  Fri, 15 Jul 2011 21:54:19 +0100

dvswitch (0.8.3.5) unstable; urgency=low

  * List all needed libraries to the linker.  We were accidentally
    depending on the behaviour of GNU ld, which treats 'needed' as
    transitive by default - Debian bug #554310

 -- Ben Hutchings <ben@decadent.org.uk>  Thu, 24 Feb 2011 04:07:10 +0000

dvswitch (0.8.3.4) unstable; urgency=low

  * Fix build failure with ffmpeg 0.6
  * Fix crash at startup with ffmpeg 0.6 due to a new parameter check

 -- Ben Hutchings <ben@decadent.org.uk>  Fri, 25 Jun 2010 12:09:12 +0100

dvswitch (0.8.3.3) unstable; urgency=low

  * Override source frame aspect ratio in case of a mismatch, rather then
    warning about it
  * Fix APTs and validity flags in header block - Alioth bug #311836

 -- Ben Hutchings <ben@decadent.org.uk>  Fri, 05 Feb 2010 13:40:10 +0100

dvswitch (0.8.3.2) unstable; urgency=low

  * Fix crash when selecting/deselecting a source that has disconnected

 -- Ben Hutchings <ben@decadent.org.uk>  Thu, 23 Jul 2009 15:12:33 +0200

dvswitch (0.8.3.1) unstable; urgency=low

  * Disable buffering of activation (tally light) notifications

 -- Ben Hutchings <ben@decadent.org.uk>  Thu, 09 Jul 2009 21:59:33 +0100

dvswitch (0.8.3) unstable; urgency=low

  * Add status overlay to make it more clear whether recording is enabled
  * Set display to black before a source is connected
  * Add support for tally lights in dvsource-firewire/dvsource-v4l2-dv,
    thanks to Wouter Verhelst

 -- Ben Hutchings <ben@decadent.org.uk>  Sun, 28 Jun 2009 20:50:02 +0100

dvswitch (0.8.2) unstable; urgency=low

  * Create one encoding thread per processor, up to a maximum of 8
  * Increase mixer queue length to allow for scheduler silliness
  * Improve robustness of dvsource-alsa
  * Fix audio encoding bug that resulted in crackling audio when using
    dvsource-alsa with NTSC

 -- Ben Hutchings <ben@decadent.org.uk>  Thu, 26 Mar 2009 22:45:15 -0500

dvswitch (0.8.1) unstable; urgency=low

  * Added dvsource-alsa as an audio-only source

 -- Ben Hutchings <ben@decadent.org.uk>  Sun, 01 Mar 2009 03:16:30 +0000

dvswitch (0.8) unstable; urgency=low

  * Added ticks and labels to the VU-meter
  * Bug fixes:
    - Removed use of function not present in Gtkmm 2.8
    - Fixed warnings from CMake 2.6
    - Fixed warnings from GCC
    - Fixed drag-selection when the pointer moves beyond the left or top
      edge of the display
    - Fixed memory corruption/crash in a selection corner-case
    - Fixed synchronisation of display buffers with the X server
    - Fixed repainting of main display before any frames have been received

 -- Ben Hutchings <ben@decadent.org.uk>  Wed, 04 Feb 2009 23:16:13 +0000

dvswitch (0.8~rc1) unstable; urgency=low

  * Disabled recording at startup
  * Added menu, currently with Quit command
  * Added buttons to GUI for effect selection
  * Changed key commands slightly
  * Allowed changing of primary and secondary video sources without
    cancelling the effect
  * Made dvsink-files create directories as necessary
  * Made dvsource-file reject non-DV files
  * Added support for USB DV sources through dvgrab and the uvcvideo driver
  * Moved source selector buttons to keep window within 1024x768 (depending
    on font selection and window decoration)
  * Added VU-meter for basic audio monitoring
  * Added warning indicator for sources which don't match the selected video
    and audio format
  * Documentation changes:
    - Documented all key commands
    - Changed contact address to dvswitch-devel@lists.alioth.debian.org
    - Added notes to (maybe) avoid user mistakes
  * Build system changes:
    - Changed to use cmake
    - Fixed building with current ffmpeg libavcodec library
  * Bug fixes:
    - Changed picture-in-picture selection to maintain frame aspect ratio
    - Restricted picture-in-picture to use the active area of the reduced
      frame
    - Slightly improved colour accuracy of picture-in-picture
    - Fixed cancellation of picture-in-picture selection
    - Fixed crash when user tries to apply picture-in-picture with an empty
      selection
    - Fixed crash when display depth is not 24 or 32
    - Fixed crash when users click in the dvswitch window without any source
      enabled
    - Made dvsource-file work with non-seekable files such as pipes

 -- Ben Hutchings <ben@decadent.org.uk>  Tue, 27 Jan 2009 01:04:43 +0000

dvswitch (0.7) unstable; urgency=low

  * Made keyboard shortcut for cut button work
  * Moved video/DV system parameters into a common structure
  * Replaced libdv codec with DV codec in libavcodec (ffmpeg), which
    is significantly faster
  * Added UI for picture-in-picture mixing

 -- Ben Hutchings <ben@decadent.org.uk>  Sun, 20 Apr 2008 22:18:37 +0100

dvswitch (0.6.2) unstable; urgency=low

  * Fixed bug in thumbnail image size initialisation that could result
    in a fatal BadValue error from X
  * Fixed minor bug in thumbnail image cleanup
  * Fixed bug that caused all sinks to be disabled when not recording

 -- Ben Hutchings <ben@decadent.org.uk>  Sun, 24 Feb 2008 10:55:41 +0100

dvswitch (0.6.1) unstable; urgency=low

  * Improved X driver compatibility:
    - Fixed calculation of Xv port numbers
    - Changed selection of visual class and depth for images to match the
      window
    - Added error messages and later checks for failure of image creation,
      where previously dvswitch could crash

 -- Ben Hutchings <ben@decadent.org.uk>  Thu, 21 Feb 2008 22:17:32 +0000

dvswitch (0.6) unstable; urgency=low

  * Added some infrastructure for video effects including
    picture-in-picture mixing
  * Switched from gtkmm 2.0/2.2 to 2.4+
  * Corrected small error in video display width
  * Changed the server to indicate overflow in output to sinks rather
    than disconnecting sinks that have overflowed; the file sink will

src/dif.c
// Copyright 2008-2009 Ben Hutchings.
// See the file "COPYING" for licence details.

#include "dif.h"

static const uint8_t dv_audio_shuffle_625_50[12][9] = {
    {   0,  36,  72,  26,  62,  98,  16,  52,  88}, /* 1st channel */
    {   6,  42,  78,  32,  68, 104,  22,  58,  94},
    {  12,  48,  84,   2,  38,  74,  28,  64, 100},
    {  18,  54,  90,   8,  44,  80,  34,  70, 106},
    {  24,  60,  96,  14,  50,  86,   4,  40,  76},
    {  30,  66, 102,  20,  56,  92,  10,  46,  82},

    {   1,  37,  73,  27,  63,  99,  17,  53,  89}, /* 2nd channel */
    {   7,  43,  79,  33,  69, 105,  23,  59,  95},
    {  13,  49,  85,   3,  39,  75,  29,  65, 101},
    {  19,  55,  91,   9,  45,  81,  35,  71, 107},
    {  25,  61,  97,  15,  51,  87,   5,  41,  77},
    {  31,  67, 103,  21,  57,  93,  11,  47,  83},
};

const struct dv_system dv_system_625_50 =
{
    .common_name = "pal",
    .frame_width = 720,
    .frame_height = 576,
    .active_region = { .left = 9, .top = 0, .right = 711, .bottom = 576 },
    .frame_rate_numer = 25,
    .frame_rate_denom = 1,
    .pixel_aspect = {
	[dv_frame_aspect_normal] = { .width =  59, .height = 54 },
	[dv_frame_aspect_wide] =   { .width = 118, .height = 81 }
    },
    .seq_count = 12,
    .size = 12 * DIF_SEQUENCE_SIZE,
    .sample_counts = {
	[dv_sample_rate_48k] =  { .min = 1896, .max = 1944, .std_cycle_len = 1, .std_cycle = { 1920 } },
	[dv_sample_rate_44k1] = { .min = 1742, .max = 1786, .std_cycle_len = 1, .std_cycle = { 1764 } },
	[dv_sample_rate_32k] =  { .min = 1264, .max = 1296, .std_cycle_len = 1, .std_cycle = { 1280 } }
    },
    .audio_shuffle = dv_audio_shuffle_625_50,
};

static const uint8_t dv_audio_shuffle_525_60[10][9] = {
    {  0, 30, 60, 20, 50, 80, 10, 40, 70 }, /* 1st channel */
    {  6, 36, 66, 26, 56, 86, 16, 46, 76 },
    { 12, 42, 72,  2, 32, 62, 22, 52, 82 },
    { 18, 48, 78,  8, 38, 68, 28, 58, 88 },
    { 24, 54, 84, 14, 44, 74,  4, 34, 64 },

    {  1, 31, 61, 21, 51, 81, 11, 41, 71 }, /* 2nd channel */
    {  7, 37, 67, 27, 57, 87, 17, 47, 77 },
    { 13, 43, 73,  3, 33, 63, 23, 53, 83 },
    { 19, 49, 79,  9, 39, 69, 29, 59, 89 },
    { 25, 55, 85, 15, 45, 75,  5, 35, 65 },
};

const struct dv_system dv_system_525_60 =
{
    .common_name = "ntsc",
    .frame_width = 720,
    .frame_height = 480,
    .active_region = { .left = 4, .top = 0, .right = 716, .bottom = 480 },
    .frame_rate_numer = 30000,
    .frame_rate_denom = 1001,
    .pixel_aspect = {
	[dv_frame_aspect_normal] = { .width = 10, .height = 11 },
	[dv_frame_aspect_wide] =   { .width = 40, .height = 33 }
    },
    .seq_count = 10,
    .size = 10 * DIF_SEQUENCE_SIZE,
    .sample_counts = {
	[dv_sample_rate_48k] = {
	    .min = 1580, .max = 1620,
	    .std_cycle_len = 5, .std_cycle = { 1602, 1601, 1602, 1601, 1602 }
	},
	[dv_sample_rate_44k1] = {
	    .min = 1452, .max = 1489,
	    .std_cycle_len = 100,
	    .std_cycle = {
		1471, 1472, 1471, 1472, 1471, 1472, 1471, 1472, 1471, 1472,
		1471, 1472, 1471, 1472, 1471, 1472, 1471, 1471, 1472, 1471,
		1472, 1471, 1472, 1471, 1472, 1471, 1472, 1471, 1472, 1471,
		1472, 1471, 1472, 1471, 1471, 1472, 1471, 1472, 1471, 1472,
		1471, 1472, 1471, 1472, 1471, 1472, 1471, 1472, 1471, 1472,
		1471, 1471, 1472, 1471, 1472, 1471, 1472, 1471, 1472, 1471,
		1472, 1471, 1472, 1471, 1472, 1471, 1471, 1472, 1471, 1472,
		1471, 1472, 1471, 1472, 1471, 1472, 1471, 1472, 1471, 1472,
		1471, 1472, 1471, 1471, 1472, 1471, 1472, 1471, 1472, 1471,
		1472, 1471, 1472, 1471, 1472, 1471, 1472, 1471, 1472, 1471
	    }
	},
	[dv_sample_rate_32k] = {
	    .min = 1053, .max = 1080,
	    .std_cycle_len = 15,
	    .std_cycle = {
		1068, 1067, 1068, 1068, 1068, 1067, 1068, 1068, 1068, 1067, 1068, 1068, 1068, 1067, 1068
	    }
	},
    },
    .audio_shuffle = dv_audio_shuffle_525_60,
};

enum dv_frame_aspect dv_buffer_get_aspect(const uint8_t * buffer)
{
    const uint8_t * vsc_pack = buffer + 5 * DIF_BLOCK_SIZE + 53;

    // If no VSC pack present, assume normal (4:3) aspect
    if (vsc_pack[0] != 0x61)
	return dv_frame_aspect_normal;

    // Check the aspect code (depends partly on the DV variant)
    int aspect = vsc_pack[2] & 7;
    int apt = buffer[4] & 7;
    if (aspect == 2 || (apt == 0 && aspect == 7))
	return dv_frame_aspect_wide;
    else
	return dv_frame_aspect_normal;
}

void dv_buffer_set_aspect(uint8_t * buffer, enum dv_frame_aspect aspect)
{
    const struct dv_system * system = dv_buffer_system(buffer);
    unsigned seq;

    for (seq = 0; seq != system->seq_count; ++seq)
    {
	uint8_t * vsc_pack =
	    buffer + seq * DIF_SEQUENCE_SIZE +
	    ((seq & 1) ? 3 * DIF_BLOCK_SIZE + 8 : 5 * DIF_BLOCK_SIZE + 53);

	vsc_pack[2] = ((vsc_pack[2] & 0xf8) |
		       ((aspect == dv_frame_aspect_wide) ? 2 : 0));
    }
}

enum dv_sample_rate dv_buffer_get_sample_rate(const uint8_t * buffer)
{
    const uint8_t * as_pack = buffer + (6 + 3 * 16) * DIF_BLOCK_SIZE + 3;

    if (as_pack[0] == 0x50)
    {
	unsigned sample_rate = (as_pack[4] >> 3) & 7;
	if (sample_rate < dv_sample_rate_count)
	    return sample_rate;
    }

    // If no AS pack present or sample rate is unrecognised, assume 48 kHz.
    // XXX Does this make any sense?
    return dv_sample_rate_48k;
}   
